pkg/kv/kv_client_lww_test.go

```diff
@@ -0,0 +1,355 @@
+// included in the /LICENSE file.
+
+package kv_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+	"github.com/stretchr/testify/require"
+)
+
+// unwrapConditionFailedError unwraps an error into a kvpb.ConditionFailedError.
+// It fails the test if there is no error or if the error cannot be unwrapped
+// into the expected type.
+func unwrapConditionFailedError(t *testing.T, err error) *kvpb.ConditionFailedError {
+	t.Helper()
+	require.Error(t, err, "expected an error")
+
+	var conditionFailedErr *kvpb.ConditionFailedError
+	require.True(t, errors.As(err, &conditionFailedErr),
+		"expected error to unwrap to *kvpb.ConditionFailedError, got: %v", err)
+
+	return conditionFailedErr
+}
+
+func TestDeleteWithOriginTimestamp(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	s, _, db := serverutils.StartServer(t, base.TestServerArgs{})
+	defer s.Stopper().Stop(context.Background())
+
+	ctx := context.Background()
+	key := roachpb.Key("lww-test-key")
+	value := []byte("value")
+
+	// Put a value with the current timestamp.
+	require.NoError(t, db.Put(ctx, key, value))
+
+	// Get the current timestamp to use as a reference.
+	gr, err := db.Get(ctx, key)
+	require.NoError(t, err)
+	currentTS := gr.Value.Timestamp
+
+	// Helper function to attempt a delete with a specific origin timestamp
+	attemptDelete := func(ts hlc.Timestamp) error {
+		b := &kv.Batch{}
+		b.Del(key)
+		b.Header.WriteOptions = &kvpb.WriteOptions{
+			OriginTimestamp: ts,
+		}
+		return db.Run(ctx, b)
+	}
+
+	t.Run("OlderTimestamp", func(t *testing.T) {
+		// Delete with an origin timestamp older than the mvcc timestamp should fail
+		err := attemptDelete(currentTS.Prev())
+		condErr := unwrapConditionFailedError(t, err)
+		require.False(t, condErr.OriginTimestampOlderThan.IsEmpty(), "expected OriginTimestampOlderThan to be set")
+		require.Equal(t, currentTS, condErr.OriginTimestampOlderThan, "expected OriginTimestampOlderThan to match current timestamp")
+
+		// Verify the original value still exists
+		gr, err := db.Get(ctx, key)
+		require.NoError(t, err)
+		require.NotNil(t, gr.Value, "expected value to exist")
+		require.Equal(t, value, gr.ValueBytes(), "value should match what was put")
+	})
+
+	t.Run("NewerTimestamp", func(t *testing.T) {
+		// Delete with an origin timestamp newer than the mvcc timestamp should succeed
+		err := attemptDelete(currentTS.Next())
+		require.NoError(t, err)
+
+		// Verify the value was deleted
+		gr, err := db.Get(ctx, key)
```

Comment:
If you wanted, you could construct a get with the option to return full mvcc value headers to verify that the value header was set.  But I think the mvcc history tests probably cover that well enough so I leave it to your judgement.

Reply:
Is that actually possible? I considered this when writing the test, but `cmd_get.go` has no way to set Tombstone: true on the MVCCGetOptions.

https://github.com/cockroachdb/cockroach/blob/8c96475bf1e5af4a90b521ab66e34e67cc7faccc/pkg/kv/kvserver/batcheval/cmd_get.go#L40

And I decided it wasn't worth the trouble to figure out how to use scan for this since the LDR tests and the mvcc tests verify the actual origin timestamp.